# CAMISIM

CAMISIM is a software to model abundance distributions of microbial communities and to simulate corresponding shotgun metagenome datasets.
It was mainly developed for the [Critical Assessment of Metagenome Annotation (CAMI)](http://microbiome-cosi.org/cami) challenge, initially by [Peter Hofmann](https://github.com/p-hofmann) and later by [Adrian Fritz](https://github.com/alphasquad), with input from members of the [CAMI team](https://github.com/orgs/CAMI-challenge/people).

### Documentation 
* [User manual](https://github.com/CAMI-challenge/CAMISIM/wiki/User-manual)
* [Software and hardware dependencies](https://github.com/CAMI-challenge/CAMISIM/wiki/Dependencies)
* [Command line Arguments](https://github.com/CAMI-challenge/CAMISIM/wiki/Command-line-Arguments)
* [Configuration File Options](https://github.com/CAMI-challenge/CAMISIM/wiki/Configuration-File-Options)
* [File Formats](https://github.com/CAMI-challenge/CAMISIM/wiki/File-Formats)

### Citation

The CAMISIM manuscript is submitted for publication, but its preprint version is available already:
* Fritz*, Hofmann*, et al. (2018). **CAMISIM: Simulating metagenomes and microbial communities.** *bioRxiv*, 300970. doi:[10.1101/300970](https://doi.org/10.1101/300970)

CAMISIM's core functionality was briefly described in the CAMI manuscript, thus you may also cite:
* Sczyrba*, Hofmann*, Belmann*, et al. (2017). **Critical Assessment of Metagenome Interpretation—a benchmark of metagenomics software.** *Nature Methods*, 14, 11:1063–1071. doi:[10.1038/nmeth.4458](https://doi.org/10.1038/nmeth.4458)
